[blackjack - Step1] Submit AnnLvu's mission

AnnLvu opened the pull request:
# Step 1 PR

Hi!  
That's my step1 PR.  
I’d like to thank you in advance for your time spent on the review and for any future advice.

This was a really interesting mission for me. I tried using many new things: an interface, parameterized tests, and lambda expressions. And I played blackjack for the first time! :) I learned so much!  
I tried to apply the MVC pattern based on my current understanding, with a strong focus on separating the responsibilities of the main controller. That led me to create multiple manager classes to handle different parts of the logic. I also aimed to follow proper class design principles by limiting access between classes and objects.  
Although TDD is still challenging to fully integrate, I'm starting to get used to the mindset. Overall, I’m happy to see clear improvement compared to my previous project in many areas.

We tried to optimize everything as much as possible, but there were still a few things we couldn't simplify. We left TODOs in two places. Maybe you could give me some advice on how I could improve them:  
- **Controller -> PlayerManager -> `fun askPlayerHit`**  
- **utils -> CardGenerator -> `fun generateSuits()`**

---

## Controller

- **GameManager:** Manages the overall game flow. Responsibilities will be split into manager classes  
- **CardManager:** Handles deck management using `ArrayDeque`, tracks used/available cards  
- **PlayerManager / DealerManager / StatsManager:** Extracted from GameManager to improve separation of concerns

## model
- **Card:** Represents a playing card with a name and a digit  
- **Hand:** Stores a list of cards and provides a method to calculate the hand’s score  
- **Rank:** Enum used for scoring hands based on card ranks  
- **Playable Interface:** Introduced for shared logic between Player and Dealer  
- **Player / Dealer:** Implementations of Playable. Dealer includes logic to decide whether to draw based on rules
-  **Stats:** Tracks individual player results and overall dealer performance 

## view

- **InputView:** Handles user inputs like player names and draw decisions. Includes validation  
- **OutputView:** Displays game state — initial hands, card draws, final results, and dealer actions

san-ghun committed:
docs(readme): add Features List

- add features for CardGenerator in new section Features List

san-ghun committed:
feat(Card,CardGenerator): add a value class and methods to generate a card

- add a value class 'Card' to hold value of suit
- add generateSuits() method which create a list of combinations for suits
- add generateCard() method which create a card instance
- add tests for new methods

san-ghun committed:
feat(CardGenerator): add 'generateCards()' method to create a list of cards

- add two test cases for the methods

san-ghun committed:
feat(InputView): add 'readPlayerNames' method and 'readYesOrNo' method

- add readPlayerNames to take user input for players names
- add readYesOrNo to take user input for 'yes or no' drawing another card

san-ghun committed:
feat(Card,Rank): update 'Card' and add 'Rank'

- change 'Card' class from value class to data class for the usage of multiple properties
- add 'Rank' enum class for later usage of calculating score of hand, which is a collection of cards
- add 'CardTest' and add a test case to validate generation of a property 'digit'

san-ghun committed:
feat(CardManager): add CardManager class with test and update README.md file

- add new class 'CardManager' as a controller unit to manage cards
- add 'checkCardMap()' method to figure out whether the card is available to take or not
- add 'giveCard()' method to get card from cards and update cardmap to check availability and return the card
- add test cases for methods in the class
- update Features List for CardManager under Controller section

san-ghun committed:
feat(Playable): add new interface 'Playable' for a character of player and dealer

san-ghun committed:
feat(Player,Rank): add data class 'Player' and update enum class 'Rank'

- data class 'Player' inherit interface 'Playable'
- add test for 'Player' in 'PlayerTest'
- delete 'BIG_ACE' from 'Rank'
- update features list in README.md file

san-ghun committed:
feat(Dealer): add data class 'Dealer'

- data class 'Dealer' inherit interface 'Playable'
- add test for 'Dealer' in 'DealerTest'
- update features list in README.md file
- remove a line that not used anymore from `PlayerTest`

san-ghun committed:
feat(GameManager): add new class 'GameManager' as a controller unit

- Working In Progress
- write rough pseudo class with comments to implement 'GameManager'
- add new section 'GameManager' in features list, README.md file

san-ghun committed:
feat(OutputView,GameManager,Application,Models): add OutputView and apply to GameManager

- add `OutputView` with methods to display the game data to user
- apply the change to `GameManager`
- add `getStringOfHand()` to `Player` and `Dealer` to make smoother interaction with `OutputView`
- add `main()` method is `Application.kt` file to run program
- update feature list in README

san-ghun committed:
feat(Stats,OutputView,GameManager): add data class 'Stats' to manage final result of game

- add 'initPlayerBoard()' to collect result of players against dealer
- add 'updateDealerStats()' to calculate and collect total score of dealer
- add 'displayFinalResult()' in 'OutputView' to display the final result based on 'Stats'
- apply these changes into 'GameManager'
- update features list in README.md file

san-ghun committed:
refactor(GameManager): tidy flow control of 'askPlayerHit()' less messy

san-ghun committed:
chore(Stats): make 'initPlayerBoard()' private

san-ghun committed:
test(StatsTest): add test for 'Stats' class

san-ghun committed:
refactor(Hand,Models,OutputView,PlayerTest): extract 'hand: List<Card>' property and logic into dedicated class 'Hand'

- extract 'hand' property and logic related with it into new data class 'Hand'
- move repeated implementation in 'Dealer' and 'Player' into interface 'Playable'
- modify 'Player', 'Dealer', 'OutputView', and 'PlayerTest' derived by the extraction of new class 'Hand'

san-ghun committed:
refactor(Stats): separate condition part from 'initPlayerBoard()'

- add new private method 'recordPlayerBoard()'

san-ghun committed:
refactor(Card,Rank,Hand): simplify 'of' method in companion object of 'Rank'

- let 'of' method in 'Rank' dynamically find and return dedicated 'Rank'
- add new property 'value' in 'Card' data class to store value of 'Rank' of 'Card'
- refactor and simplify 'calculateCards()' method in 'Hand' data class

san-ghun committed:
refactor(Stats): add constants to improve readability of game result

- use companion object to assign constant values for win/lose/tie

san-ghun committed:
refactor(OutputView): add new method for refactoring and add constants to improve readability of player results

- extract conditional part of 'displayFinalResults()' into new private method 'givePlayerResult()'
- use private object to assign constant values for win/lose/tie

san-ghun committed:
fix(InputView): add input validation for blank input with 'require()'

- add TODO comment in GameManager for later

san-ghun committed:
test(HandTest): add test for 'Hand' data class

san-ghun committed:
feat(Controller): add 'PlayerManager' as a controller unit and apply to 'GameManager'

- extract 'Player' related parts from 'GameManager' to new class 'PlayerManager'

san-ghun committed:
feat(Controller): add 'DealerManager' as a controller unit and apply to 'GameManager'

san-ghun committed:
feat(Controller): add 'StatsManager' as a controller unit and apply to 'GameManager'

- extract statistics related parts from 'GameManager' to 'StatsManager'

san-ghun committed:
feat(InputView): add new method 'retryable()' to re-prompt user input for invalid input

san-ghun committed:
refactor(Controllers,Application): extract view parts from 'PlayerManager' and 'StatsManager' to 'GameManager'

- change 'GameManger' from class to object to make it singleton unit
- split 'run()' method into parts to separate responsibilities into parts, in 'GameManager'
- add 'InputView' and 'OutputView' as properties in 'GameManger' to have access to views
- in 'PlayerManager' add two methods to call 'GameManager' and access to view methods

san-ghun committed:
refactor(OutputView): update 'displayInitialState()' method

- change string formation process for 'displayInitialState()' method
- remove some comments in 'Dealer'

san-ghun committed:
refactor(PlayerManager,StatsManager): cover open properties and methods into private

san-ghun committed:
fix(InputView): add input validation for duplicates in names for 'readPlayerNames()'

san-ghun committed:
docs(readme): mark and update features list in README.md file

san-ghun committed:
chore: remove resolved TODO comment

Hyeon9mak commented on README.md:
```diff
@@ -1 +1,115 @@
+
+## Features List
+
+### Model
+
+Card
+
+- [x] Card class hold string value of name(e.g. "A♥") and string value of digit(e.g. "A", "5", "K")
+
+Hand
+
+- [x] has `List<Card>` for player and dealer
+- [x] extract logic for calculation of hand's score from `Player` and `Dealer` into `calculateCards()`
+
+Rank
+
+- [x] Add new enum class 'Rank'(for example) and use the class to calculate score of player's hand and dealer's hand
+
+Playable (interface)
+
+- [x] has value `name` in string
+- [x] has list of card called `hand`
+- [x] has method `requestCard()` to request a card by condition then, ask the card manager with boolean
+- [x] has method `drawCard()` to take card given by a card manager
+- [x] has method `calculateHand()` to calculate score of hand
+- [x] has method `isBust()` to figure out the player of the dealer is bust or not
+
+Player: Playable
+
+- [x] follows interface `Playable`
+- [x] has method `calculateHand()` to calculate score of hand with `Rank` of `Card`
+    - TODO: MUST have to find a solid logic for calculation
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/controller/CardManager.kt:
```diff
@@ -0,0 +1,46 @@
+        deque.removeFirst()
+        _cards = deque.toList()
+        return card
+    }
```
I think the Controller's role in the MVC pattern is to help communication between Model(Domain) and View.
That means all business rules should be baked into the Model(Domain), so when business policies change, fellow developers only need to look at the Model(Domain) package.

Since the Controller helps communication between Model and View,
 I think it can naturally manage the business flow to some extent. 

But the current structure has the Controller pulling out all properties from the Model and handling everything one by one.
I'm guessing writing test code was pretty challenging, wasn't it? 🤔

Hyeon9mak replied:
How about trying to let the Model classes interact with each other more directly?
Then naturally, the need for Manager classes will gradually disappear. 😉 

```kotlin
dealer.selfDrawInitCards()
dealer.giveInitCardsToPlayers(players = players)
```

AnnLvu replied:
Changed it! Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/controller/PlayerManager.kt:
```diff
@@ -0,0 +1,48 @@
+                    callGameManagerToReadYesOrNo(player.name)
+                }
+            if (requestMessage) {
+                isFirst = false
```
How about make negative asking method?
It can be remove `!` operator. `!` operator make lot of mistake.

AnnLvu replied:
Create new method fun canContinue(): Boolean

Hyeon9mak commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,6 @@
+package blackjack.model
+
+data class Card(val name: String) {
+    val digit = name.substring(0, name.length - 1)
+    val value = Rank.of(digit).value
+}
```
I guess u want like this...

```kotlin
data class Card(name: String) {
    // ...
}
```

but, Does Card really need to know about strings? 🤔

AnnLvu replied:
You're right to ask! I kept Card(rank, suit) to avoid string parsing logic inside the model

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,22 @@
+package blackjack.model
+
+data class Dealer(override val name: String = "Dealer") : Playable {
+    private var _hand = Hand()
+    override val hand: Hand
+        get() = _hand
```
If the name "Dealer" needs to be changed to below image file, which class file will a fellow developer open first: `OutputView`-`Dealer`?

<img width="862" height="534" alt="image" src="https://github.com/user-attachments/assets/f0acb4f5-574d-46e2-a121-ac3b36268fb5" />

AnnLvu replied:
Previously, to change the dealer's name, developer had to modify OutputView manually.
Now, the dealer’s display name is handled directly inside the Dealer class, so developers only need to update Dealer - no need to touch OutputView anymore.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,22 @@
+package blackjack.model
+
+data class Hand(val cards: List<Card> = emptyList()) {
```
What's different between data class & value class?
- https://kotlinlang.org/docs/inline-classes.html
- https://kotlinlang.org/docs/data-classes.html

AnnLvu replied:
Thanks for the links and question!
I kept `Hand` as a `data class` because it holds a `List<Card>` and relies on auto-generated `toString()` for terminal output in `OutputView`. The `calculateCards()` method also adds behavior, making `Hand` a complex entity.

Hyeon9mak replied:
Doesn't it the `value class` also auto-generate `toString()`? 🤔

AnnLvu replied:
Hi! Thanks for the comment and question about value class and toString()!  I decided to keep Hand as a data class and retain the toText() method for the following reasons:
- Hand is not just a wrapper around List<Card> but an entity with additional logic (calculateCards() for scoring with ace handling). A data class is better suited for such cases, as it supports complex data structures and auto-generates useful methods
- The toText() method provides a clean, user-friendly output (A♥, K♠) that’s perfect for the game’s console interface. The auto-generated toString() from a data class outputs Hand(cards=[A♥, K♠]) - in our Example Output we have (A♥, K♠)

Hyeon9mak commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,22 @@
+    fun calculateCards(): Int {
+        val values = cards.map { it.value }
+        var aceCounts = values.count { it == 11 }
+        var score = values.sum()
+        while (aceCounts > 0 && score > 21) {
+            score -= 10
+            aceCounts--
+        }
+        if (score < 10 && values.contains(11)) {
+            score += 10
+        }
+        return score
```
I think u can give a new name for 21 score. 😄

AnnLvu replied:
Create a new object Constants. Thank you!

Hyeon9mak commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -0,0 +1,18 @@
+package blackjack.model
+
+interface Playable {
+    val name: String
+    val hand: Hand
```
What's difference between interface & abstract class?

AnnLvu replied:
Yeah, originally Playable was an interface because it felt like `Player` and `Dealer` were just different participants in the game, each with potentially unique behavior.
Player and Dealer actually share a lot of the same logic .So it's not just similar structure - they literally do the same things in many places.
That’s why I changed `Playable` from an interface to an abstract class. It allowed me to centralize shared behavior, reduce duplication, and make the architecture cleaner and easier to maintain.
An **interface** is more about what something can do, but doesn't say how. Interfaces are great when you only need to define a contract.
An **abstract class** lets you define both what and how, which fits our case better.

Hyeon9mak commented on src/main/kotlin/blackjack/model/Rank.kt:
```diff
@@ -0,0 +1,25 @@
+    KING("K", 10),
+    NONE("0", 0),
+    ;
+
```
When the Blackjack app convert to Web browser image game,
fellow develop should check the `InputView`, `OutputView`, `Card`, ... a lot of classes.

How about move all view logic to View classes?
Or, u can make duplicate enum class in View class file only used for output on terminal.

```kotlin
enum class SuitView(
    val suit: Suit,
    val symbol: String,
) {
    HEART(suit = Suit.HEART, symbol = "♥"),
    // ...
}
```

AnnLvu replied:
Thanks for the suggestions! I've added `SuitView` and `RankView` in `blackjack.view` for display logic

Hyeon9mak commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -0,0 +1,48 @@
+package blackjack.model
+
+data class Stats(val players: List<Player>, val dealer: Dealer) {
+    private var _playerBoard = initPlayerBoard()
+    val playerBoard: Map<Player, Int> get() = _playerBoard
+
+    private var _dealerStats = mapOf("win" to 0, "lose" to 0, "tie" to 0)
+    val dealerStats: Map<String, Int> get() = _dealerStats
+
```
I think those are View's responsibility. 😮 
How do u think?

AnnLvu replied:
I moved it to the View folder and renamed it to StatsView!

Hyeon9mak commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,34 @@
+        return trimmedInput
+    }
+
+    fun readYesOrNo(name: String): Boolean {
+        println("\nWould $name like to draw another card? (y for yes, n for no)")
+        val input = readln().trim()
+        return when (input) {
+            "y" -> true
+            "n" -> false
+            else -> throw IllegalArgumentException("Please write 'y' for 'yes' or 'n' for 'no'.")
+        }
+    }
```
Yeah! `y`, `n` strings are not important things of Blackjack game.
Those are just user's input buttons. 💯 👍

but.. 😱 

<img width="439" height="128" alt="image" src="https://github.com/user-attachments/assets/e61bf6ba-cd82-4dea-8bc9-0b0a4ccd10eb" />

AnnLvu replied:
I’ve changed it -now, every time the user enters an incorrect "y" or "n", there’s an opportunity to enter the answer again right away, and only then does the next message appear.

Hyeon9mak reviewed (changes requested):
Hello Ann!
Nice work to complete Blackjack mission step1 👍

I was really impressed by your attempt to split the Controller into multiple parts to distribute responsibilities. 
Since HeroTechCourse is all about trying, experiencing, and improving different ways to write maintainable code, I want to encourage these kinds of attempts.

Unfortunately, through this experience, you probably noticed that while Controller responsibilities were distributed, the Model (Domain) classes became too lightweight as a downside. 
To solve this problem, I think we need to change the approach from pulling properties out of Model (Domain) classes to handle them in Controllers, to having Model (Domain) classes communicate directly with each other.

I've left some related comments, so please take your time checking them out and feel free to ask questions if you run into any difficulties.

Feel free to ask any questions u have! 😉

AnnLvu committed:
refactor(controller, model): move initial card distribution logic into Dealer
- Rename `CardManager` to `Deck` for clearer domain terminology
- Rename `giveCard` to `drawCard`
- Create `Dealer.giveInitCardsToPlayers(players, deck)` to distribute cards to all players
- Create `Dealer.selfDrawInitCards(deck)` to draw two cards for the dealer
- Remove the now-unnecessary `initHands()` function from GameManager

AnnLvu committed:
refactor(controller, model): replace !player.isBust() with player.canContinue() to improve readability
- add new method `canContinue()` into interface `Playable`

AnnLvu committed:
refactor(blackjack): remove string-based Card constructor in favor of typed Rank & Suit

- Replace `Card(name: String)` with `Card(rank: Rank, suit: Suit)` to improve type safety
- Create Suit enum to represent card suits clearly
  and decouple the Card model from string parsing logic
- change tests

AnnLvu committed:
refactor(view, controller): replace hardcoded "Dealer" with dealer.name in OutputView

AnnLvu committed:
feat(constants): extract magic numbers to Constants.kt

AnnLvu committed:
refactor(model): convert Playable from interface to abstract class
- Switch `Playable` from interface to abstract class to consolidate shared logic
  between Player and Dealer, including drawCard, calculateHand, isBust, and
  canContinue.

AnnLvu committed:
feat(view): move Rank and Suit symbols to view layer

AnnLvu committed:
refactor(view): move Stats to view layer and renamed to StatsView for clearer separation of concerns

AnnLvu committed:
fix(stats): correct final statistics output and "y" or "n" message

Hyeon9mak commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -0,0 +1,32 @@
+        return calculateHand() > Constants.BLACK_JACK
+    }
+
+    fun canContinue(): Boolean {
+        return calculateHand() <= Constants.BLACK_JACK
+    }
```
😉 👍

Hyeon9mak commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,27 @@
+package blackjack.model
+
+import blackjack.controller.Deck
+
+data class Dealer(override val name: String = "Dealer") : Playable() {
+    fun shouldDrawCardOrNot(): Boolean {
+        val score = calculateHand()
+        return score <= 16
+    }
+
```
What do you think is the role of the View in the MVC pattern?
Could you explain how you see the responsibilities being divided between the Model and the View?
I think that would help me better understand your intent. 💭

Hyeon9mak commented on src/main/kotlin/blackjack/model/Playable.kt:
```diff
@@ -0,0 +1,32 @@
+
+import blackjack.utils.Constants
+
+abstract class Playable {
+    abstract val name: String
+    protected var handInternal: Hand = Hand()
+    open val hand: Hand
+        get() = handInternal
+
+    open fun requestCard(condition: () -> Boolean): Boolean {
+        return condition()
+    }
+
+    open fun drawCard(newCard: Card) {
+        val deque = ArrayDeque(hand.cards)
+        deque.addLast(newCard)
+        handInternal = Hand(deque.toList())
+    }
```
To put it more intuitively, you can decide between using an interface or an abstract class based on whether the rules of the real-world concept are likely to change.

For example, the rules of Blackjack have been passed down for over a hundred years and are unlikely to change anytime soon. Based on the assumption that "both the dealer and the player are participants," it's reasonable to use an abstract class.

On the other hand, if a component is difficult to predict in terms of how it might change or expand, it's better to use an interface to leave room for flexibility and future extensions.

AnnLvu replied:
Thank you so much for explanation!

Hyeon9mak commented on src/main/kotlin/blackjack/view/SuitView.kt:
```diff
@@ -0,0 +1,23 @@
+            Suit.NONE -> "?"
+        }
+
+    fun fromSymbol(symbol: Char): Suit =
+        when (symbol) {
+            '♥' -> Suit.HEARTS
+            '♦' -> Suit.DIAMONDS
+            '♣' -> Suit.CLUBS
+            '♠' -> Suit.SPADES
+            else -> Suit.NONE
+        }
```
I think it's only used in test code 🤔

Hyeon9mak commented on src/main/kotlin/blackjack/controller/PlayerManager.kt:
```diff
@@ -0,0 +1,48 @@
+        _players = mutableList.toList()
+    }
+
+    // TODO: resolve indent depth
+    fun askPlayerHit(
+        player: Player,
+        receiveCard: () -> Card,
```
need double check the TODO comments!

AnnLvu replied:
OMG, I forgot about this, sorry!

Hyeon9mak commented on src/main/kotlin/blackjack/Application.kt:
```diff
@@ -0,0 +1,8 @@
+
+import blackjack.controller.GameManager
+
+fun main() {
+    val gm = GameManager
+    gm.run()
+}
```
<img width="558" height="344" alt="image" src="https://github.com/user-attachments/assets/1e152f01-3366-4e46-b67a-d364583311f1" />

AnnLvu replied:
Rename dm -> gameManager

Hyeon9mak commented on src/main/kotlin/blackjack/controller/DealerManager.kt:
```diff
@@ -0,0 +1,7 @@
+
+import blackjack.model.Dealer
+
+class DealerManager {
+    val dealer = Dealer()
+}
```
is this really need class? 🤔

AnnLvu replied:
I suppose, no. I delete DealerManager and use Dealer directly in GameManager

Hyeon9mak reviewed (approved):
Hi Ann,
Great work incorporating the feedback from step1!

There are a few comments where I left follow-up questions because I wasn’t fully sure about your intentions. 
When you submit the PR for step2, it would be great if you could also include your answers to those questions.
— that way, I can give you more specific feedback.

And of course, if you have any questions yourself, feel free to ask anytime!

Enjoy the rest of your weekend! 😊
